app/codescanner.cpp

```diff
@@ -1,159 +1,81 @@
 #include "codescanner.h"
 
+#include <QTimer>
+
+#include "imageutils.h"
+
 CodeScanner::CodeScanner( QObject *parent )
-  : QVideoSink( parent )
-  , mCamera( nullptr )
+  : QObject( parent )
 {
-  connect( &mDecoder, &QRDecoder::capturedStringChanged, this, &CodeScanner::setCapturedString );
-  connect( this, &QVideoSink::videoFrameChanged, this, &CodeScanner::processFrame );
-
   mWorker = new QRWorker( this );
   mWorker->moveToThread( &mWorkThread );
+
+  connect( this, &CodeScanner::startProcessing, mWorker, &QRWorker::process );
+  connect( mWorker, &QRWorker::codeScanned, this, &CodeScanner::codeScanned );
   connect( &mWorkThread, &QThread::finished, mWorker, &QObject::deleteLater );
-  connect( this, &CodeScanner::process, mWorker, &QRWorker::process );
-  mWorkThread.start();
 
-  initCamera();
+  mWorkThread.start();
 }
 
 CodeScanner::~CodeScanner()
 {
   mWorkThread.quit();
   mWorkThread.wait();
-  stopCamera();
-}
-
-void CodeScanner::initCamera()
-{
-  mCamera = new QCamera( this );
-  const auto settings = mCamera->cameraDevice().videoFormats();
-
-#ifdef Q_OS_ANDROID
-  int i = mCamera->cameraDevice().videoFormats().size() - 1;
-#else
-  int i = 0;
-#endif
-
-  const auto s = settings.at( i );
-
-  int w = settings.at( i ).resolution().width();
-  int h = settings.at( i ).resolution().height();
-  mDecoder.setResolution( w, h );
-
-  mCamera->setFocusMode( QCamera::FocusModeAuto );
-  mCamera->setExposureMode( QCamera::ExposureBarcode );
-  mCamera->setCameraFormat( s );
-
-  mCaptureSession.setCamera( mCamera );
-  mCaptureSession.setVideoSink( this );
-
-  mCamera->start();
-}
-
-void CodeScanner::stopCamera()
-{
-  mCamera->stop();
-  disconnect( mCamera, nullptr, nullptr, nullptr );
-  mCamera->setParent( nullptr );
-  delete mCamera;
-  mCamera = nullptr;
 }
 
 void CodeScanner::processFrame( const QVideoFrame &frame )
 {
-  if ( mProcessing )
+  if ( mIgnoreFrames )
   {
-    emit process( mDecoder.videoFrameToImage( frame, captureRect().toRect() ) );
-
-    if ( mVideoSink )
-    {
-      mVideoSink->setVideoFrame( frame );
-    }
+    return;
   }
-  pauseProcessing();
-}
 
-void CodeScanner::processImage( QRDecoder *decoder, const QImage &image )
-{
-  decoder->processImage( image );
-  continueProcessing();
-}
+  QImage image = ImageUtils::rescaledImage( frame );
```

Comment:
Can we have a check for image validity here? In some cases `rescaledImage` returns empty image and there is no sense in trying to process it.